### ISSUE TITLE
[MODULES-136] When JBoss Modules are used as a library causes FactoryConfigurationError

### ISSUE THREAD

**Issue by dmlloyd** [jboss-modules/jboss-modules#434]
**[Fred Bricon](https://issues.redhat.com/secure/ViewProfile.jspa?name=fbricon)** opened **[MODULES-136](https://issues.redhat.com/browse/MODULES-136?redirect=false)** and commented

Issue was original about use of jboss modules in Eclipse via m2e-apt/jboss logging but is also causing problems for forge and arquillian.

Original:
JBDS 5.5.0.CR1 embeds [m2e-apt](https://github.com/jbosstools/m2e-apt), which automatically enables Eclipse JDT Annotation Processor Toolkit.
When importing the [server](https://github.com/jbossas/jboss-as/tree/master/server) module from http://github.com/jbossas/jboss-as.git (also happens for other modules), Eclipse starts crashing (see JB[IDE-12087](https://issues.redhat.com/browse/IDE-12087)) with errors like :

```
javax.xml.parsers.FactoryConfigurationError: Provider __redirected.__DocumentBuilderFactory not found
	at javax.xml.parsers.DocumentBuilderFactory.newInstance(DocumentBuilderFactory.java:127)
	at org.eclipse.jdt.apt.core.internal.util.FactoryPathUtil.decodeFactoryPath(FactoryPathUtil.java:222)
	at org.eclipse.jdt.apt.core.internal.util.FactoryPathUtil.readFactoryPathFile(FactoryPathUtil.java:115)
	at org.eclipse.jdt.apt.core.internal.util.FactoryPathUtil.calculatePath(FactoryPathUtil.java:342)
	at org.eclipse.jdt.apt.core.internal.util.FactoryPathUtil.getFactoryPath(FactoryPathUtil.java:429)
	at org.eclipse.jdt.apt.core.internal.AnnotationProcessorFactoryLoader.getJava6FactoriesAndAttributesForProject(AnnotationProcessorFactoryLoader.java:420)
	at org.eclipse.jdt.internal.apt.pluggable.core.dispatch.IdeAnnotationProcessorManager.processAnnotations(IdeAnnotationProcessorManager.java:130)
	at org.eclipse.jdt.internal.compiler.Compiler.processAnnotations(Compiler.java:813)
	at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:432)
	at org.eclipse.jdt.internal.core.builder.AbstractImageBuilder.compile(AbstractImageBuilder.java:364)
	at org.eclipse.jdt.internal.core.builder.BatchImageBuilder.compile(BatchImageBuilder.java:178)
	at org.eclipse.jdt.internal.core.builder.AbstractImageBuilder.compile(AbstractImageBuilder.java:301)
	at org.eclipse.jdt.internal.core.builder.BatchImageBuilder.build(BatchImageBuilder.java:60)
	at org.eclipse.jdt.internal.core.builder.JavaBuilder.buildAll(JavaBuilder.java:254)
	at org.eclipse.jdt.internal.core.builder.JavaBuilder.build(JavaBuilder.java:173)
	at org.eclipse.core.internal.events.BuildManager$2.run(BuildManager.java:728)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:199)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:239)
	at org.eclipse.core.internal.events.BuildManager$1.run(BuildManager.java:292)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:42)
	at org.eclipse.core.internal.events.BuildManager.basicBuild(BuildManager.java:295)
	at org.eclipse.core.internal.events.BuildManager.basicBuildLoop(BuildManager.java:351)
	at org.eclipse.core.internal.events.BuildManager.build(BuildManager.java:374)
	at org.eclipse.core.internal.events.AutoBuildJob.doBuild(AutoBuildJob.java:143)
	at org.eclipse.core.internal.events.AutoBuildJob.run(AutoBuildJob.java:241)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:54)
Caused by: java.lang.ClassNotFoundException: __redirected/__DocumentBuilderFactory
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:264)
	at javax.xml.parsers.FactoryFinder.getProviderClass(FactoryFinder.java:119)
	at javax.xml.parsers.FactoryFinder.newInstance(FactoryFinder.java:144)
	at javax.xml.parsers.FactoryFinder.find(FactoryFinder.java:185)
	at javax.xml.parsers.DocumentBuilderFactory.newInstance(DocumentBuilderFactory.java:121)
```

It turns out the Eclipse System Properties have been changed to use javax.xml.parsers.SAXParserFactory=__redirected/__DocumentBuilderFactory

I tracked the code responsible with messing with the properties :

```
Thread [Worker-24] (Suspended (breakpoint at line 779 in System))	
System.setProperty(String, String) line: 779	
__SAXParserFactory.<clinit>() line: 68	
__JAXPRedirected.initAll() line: 80	
Module$1.run() line: 126	
Module$1.run() line: 113	
AccessController.doPrivileged(PrivilegedAction<T>) line: not available [native method]	
Module.<clinit>() line: 113	
Class<T>.forName0(String, boolean, ClassLoader) line: not available [native method]	
Class<T>.forName(String) line: 186	<====== org.jboss.modules.Module
JCodeModel.ref(String) line: 363	
MessageBundleImplementor(ImplementationClassModel).createBundleMethod(MessageMethod, JMethod, JMethod, JVar) line: 150	
MessageBundleImplementor.generateModel() line: 86	
MessageBundleImplementor(ClassModel).create(JavaFileObject) line: 104	
ImplementationClassGenerator.processTypeElement(TypeElement, TypeElement, MessageInterface) line: 63	
LoggingToolsProcessor.process(Set<TypeElement>, RoundEnvironment) line: 155	
RoundDispatcher.handleProcessor(ProcessorInfo) line: 139	
RoundDispatcher.round() line: 121	
IdeAnnotationProcessorManager(BaseAnnotationProcessorManager).processAnnotations(CompilationUnitDeclaration[], ReferenceBinding[], boolean) line: 159	
IdeAnnotationProcessorManager.processAnnotations(CompilationUnitDeclaration[], ReferenceBinding[], boolean) line: 134	
Compiler.processAnnotations() line: 813	
Compiler.compile(ICompilationUnit[]) line: 432	
BatchImageBuilder(AbstractImageBuilder).compile(SourceFile[], SourceFile[], boolean) line: 364	
BatchImageBuilder.compile(SourceFile[], SourceFile[], boolean) line: 178	
BatchImageBuilder(AbstractImageBuilder).compile(SourceFile[]) line: 301	
BatchImageBuilder.build() line: 60	
JavaBuilder.buildAll() line: 254	
JavaBuilder.build(int, Map, IProgressMonitor) line: 173	
BuildManager$2.run() line: 728	
SafeRunner.run(ISafeRunnable) line: 42	
BuildManager.basicBuild(int, IncrementalProjectBuilder, Map<String,String>, MultiStatus, IProgressMonitor) line: 199	
BuildManager.basicBuild(IBuildConfiguration, int, IBuildContext, ICommand[], MultiStatus, IProgressMonitor) line: 239	
BuildManager$1.run() line: 292	
SafeRunner.run(ISafeRunnable) line: 42	
BuildManager.basicBuild(IBuildConfiguration, int, IBuildContext, MultiStatus, IProgressMonitor) line: 295	
BuildManager.basicBuildLoop(IBuildConfiguration[], IBuildConfiguration[], int, MultiStatus, IProgressMonitor) line: 351	
BuildManager.build(IBuildConfiguration[], IBuildConfiguration[], int, IProgressMonitor) line: 374	
AutoBuildJob.doBuild(IProgressMonitor) line: 143	
AutoBuildJob.run(IProgressMonitor) line: 241	
Worker.run() line: 54	
```

That stacktrace represents the processing of https://github.com/jbossas/jboss-as/blob/master/server/src/main/java/org/jboss/as/server/ServerMessages.java#L507

So basically, whenever a Logger has a dependency to the [org.jboss.modules.Module](https://github.com/jbossas/jboss-modules/blob/master/src/main/java/org/jboss/modules/Module.java) class, stuff starts hitting the fan.

Ideally we should process the classes in an isolated way, so that doesn't break IDE embdeding.

Currently, JBoss AS source code can only be used in JBDS if m2e-apt is disabled (Preferences > Maven > Annotation Processing).

---

**Attachments:**
- [jboss-logging-generator-bug.zip](https://issues.redhat.com/secure/attachment/12354720/jboss-logging-generator-bug.zip) (_10.74 kB_)

**Issue Links:**
- [FORGE-820](https://issues.redhat.com/browse/FORGE-820) No XPathFactory implementation found inside a forge addon
 (_**"blocks"**_)
- [JBIDE-12087](https://issues.redhat.com/browse/JBIDE-12087) JDS5 beta 3 crashes when importing JBoss AS 7.1.1.Final Maven project
 (_**"blocks"**_)
- [LOGTOOL-51](https://issues.redhat.com/browse/LOGTOOL-51) Logging generation processing breaks APT rules by loading classes via reflection
 (_**"clones"**_)

-1 votes, 11 watchers

**Comment by dmlloyd** [comment]
**[Fred Bricon](https://issues.redhat.com/secure/ViewProfile.jspa?name=fbricon)** commented

Attaching simple test project showcasing the issue with eclipse + m2e-apt

**Comment by dmlloyd** [comment]
@dmlloyd commented

This is caused by the automatic registration by jboss-modules of all of the JAXP hooks which should really only happen when running from the command line.

**Comment by dmlloyd** [comment]
**[Max Andersen](https://issues.redhat.com/secure/ViewProfile.jspa?name=maxandersen)** commented

yes, but why is the annotation processor even trying to load user classes ? APT processors should not do that.

**Comment by dmlloyd** [comment]
@dmlloyd commented

That's a separate issue.

**Comment by dmlloyd** [comment]
**[Max Andersen](https://issues.redhat.com/secure/ViewProfile.jspa?name=maxandersen)** commented

But that is the root cause - this means the APT processor is breaking the APT contracts afaics.

And as a consequence cannot be run safely within a APT compile phase since it can trigger user code (such as jboss modules being bad in its init code).

**Comment by dmlloyd** [comment]
**[Max Andersen](https://issues.redhat.com/secure/ViewProfile.jspa?name=maxandersen)** commented

i.e. today jboss modules messes it up, tomorrow some user class calls something else that messes up things.

Which is why we reported it in logging module.

Looking at the code the culprit is the use of com.sun.codemodel.internal.* which definitely can cause problems when used from inside APT.

**Comment by dmlloyd** [comment]
@dmlloyd commented

I take this issue as reported to be about the Module class init installing XML handlers when that should probably be done in Main.  If you want to open a new issue about the logging tools using reflection instead of the mirror API, that's fine.

**Comment by dmlloyd** [comment]
@dmlloyd commented

Jason mentioned that he has some ideas to mitigate this.

**Comment by dmlloyd** [comment]
**[Lincoln Baxter III](https://issues.redhat.com/secure/ViewProfile.jspa?name=lincolnthree)** commented

Is there a workaround for this that doesn't require Forking and modifying Module.class's static init method?

If I'm assuming correctly that these classes are included in order to support module.xml files

Since moving the classes to a new package doesn't work, is it possible to provide a modules JAR that does not bundle the javax.xml stuff, with the extras for using module.xml repositories in a separate JAR, like so?

jboss-modules-core
jboss-modules (has the Main[] and module.xml stuff)

This is currently blocking us from running Forge directly in Eclipse.

**Comment by dmlloyd** [comment]
**[Lincoln Baxter III](https://issues.redhat.com/secure/ViewProfile.jspa?name=lincolnthree)** commented

Scratch my suggestion, David educated me; however, is is still a problem for Forge. I've forked JBoss Modules and removed this functionality to work around this issue in the mean time.

Link to fork:
https://github.com/forge/jboss-modules

**Comment by dmlloyd** [comment]
@n1hility commented

Max pinged me last night to look at this. However I don't quite follow how Lincoln's problem relates. The logging stuff was fixed a long long time ago.

**Comment by dmlloyd** [comment]
@dmlloyd commented

The fix for this (very convoluted) issue is simply to eliminate the static init stuff from the various modules classes and use explicit initialization instead, so that you can include `jboss-modules.jar` and reference its classes without suddenly installing URL handlers and loggers and XML parsers and stuff.

**Comment by dmlloyd** [comment]
**[Max Andersen](https://issues.redhat.com/secure/ViewProfile.jspa?name=maxandersen)** commented

I've updated the description since it is not limited to the APT logging problem in Eclipse.

It affects all libraries that uses jboss modules as a library nested inside other non-jboss modules systems.

**Comment by dmlloyd** [comment]
**[Lincoln Baxter III](https://issues.redhat.com/secure/ViewProfile.jspa?name=lincolnthree)** commented

My problem, like the original report, involves running Modules/Forge in JBoss Dev Studio/OSGi.

**Comment by dmlloyd** [comment]
**[George Gastaldi](https://issues.redhat.com/secure/ViewProfile.jspa?name=gastaldi)** commented

Any evolution on this issue? Thanks.

**Comment by dmlloyd** [comment]
**[Lincoln Baxter III](https://issues.redhat.com/secure/ViewProfile.jspa?name=lincolnthree)** commented

Hey guys, this is still a critical issue for us since in order to work around it and get Forge to run in eclipse at all we had to disable the Factory configuration in a private fork - https://github.com/forge/jboss-modules - which in turn means that XPath and such don't work. See FORGE-820 for more information on this.

**Comment by dmlloyd** [comment]
@dmlloyd commented

Well, I **think** this is fixed now, hopefully without messing up the init order which is still a bit too fragile for my liking.  The initialization of the VM-ownership stuff like JAXP redirection is now part of Main.main(), so embedding Modules should bypass it as long as the __redirected classes are not referenced manually or something.

**Comment by dmlloyd** [comment]
@dmlloyd commented

Unfortunately this fix has to be reverted, because it is causing other critical issues with WildFly.  We need to come up with some new approach.

**Comment by dmlloyd** [comment]
**[Lincoln Baxter III](https://issues.redhat.com/secure/ViewProfile.jspa?name=lincolnthree)** commented

So what about including the system property that I suggested? It would at least provide a pre-classloading toggle for disabling this feature entirely.

**Comment by dmlloyd** [comment]
@fromage commented

This `__redirected` static init has caused us nothing but trouble in Ceylon. I really wish it would go away (I don't understand what it does, so I assume (and hope) it's specific to WildFly and that I don't need it).

Today's Java 9 experiment also pointed to it:

```
Exception in thread "main" java.lang.IllegalAccessError: class __redirected.__SAXParserFactory cannot access class com.sun.org.apache.xerces.internal.jaxp.SAXParserFactoryImpl (in module java.xml) because module java.xml does not export package com.sun.org.apache.xerces.internal.jaxp to <unnamed module @185d8b6>
	at sun.reflect.Reflection.throwIllegalAccessException(java.base@9.0/Reflection.java:455)
	at sun.reflect.Reflection.ensureMemberAccess(java.base@9.0/Reflection.java:130)
	at java.lang.reflect.AccessibleObject.slowCheckMemberAccess(java.base@9.0/AccessibleObject.java:370)
	at java.lang.reflect.AccessibleObject.checkAccess(java.base@9.0/AccessibleObject.java:362)
	at java.lang.reflect.Constructor.newInstance(java.base@9.0/Constructor.java:435)
	at __redirected.__SAXParserFactory.<init>(__SAXParserFactory.java:106)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(java.base@9.0/Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(java.base@9.0/NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(java.base@9.0/DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(java.base@9.0/Constructor.java:444)
	at java.lang.Class.newInstance(java.base@9.0/Class.java:528)
	at javax.xml.parsers.FactoryFinder.newInstance(java.xml@9.0/FactoryFinder.java:193)
	at javax.xml.parsers.FactoryFinder.newInstance(java.xml@9.0/FactoryFinder.java:152)
	at javax.xml.parsers.FactoryFinder.find(java.xml@9.0/FactoryFinder.java:233)
	at javax.xml.parsers.SAXParserFactory.newInstance(java.xml@9.0/SAXParserFactory.java:132)
	at com.redhat.ceylon.cmr.api.Overrides.parseXml(Overrides.java:465)
	at com.redhat.ceylon.cmr.api.Overrides.parse(Overrides.java:261)
	at com.redhat.ceylon.cmr.api.Overrides.parseDistOverrides(Overrides.java:252)
	at com.redhat.ceylon.cmr.api.Overrides.getDistOverrides(Overrides.java:91)
	at com.redhat.ceylon.cmr.impl.AbstractNodeRepositoryManager.<init>(AbstractNodeRepositoryManager.java:68)
	at com.redhat.ceylon.cmr.impl.RootRepositoryManager.<init>(RootRepositoryManager.java:60)
	at com.redhat.ceylon.cmr.impl.RepositoryManagerBuilderImpl.<init>(RepositoryManagerBuilderImpl.java:57)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(java.base@9.0/Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(java.base@9.0/NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(java.base@9.0/DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(java.base@9.0/Constructor.java:444)
	at com.redhat.ceylon.cmr.api.RepositoryManagerBuilder.<init>(RepositoryManagerBuilder.java:75)
	at com.redhat.ceylon.cmr.ceylon.CeylonUtils$CeylonRepoManagerBuilder.buildManagerBuilder(CeylonUtils.java:359)
	at ceylon.modules.jboss.runtime.AbstractJBossRuntime.createRepository(AbstractJBossRuntime.java:122)
	at ceylon.modules.jboss.runtime.AbstractJBossRuntime.createRepository(AbstractJBossRuntime.java:146)
	at ceylon.modules.jboss.runtime.JBossRuntime.createModuleLoader(JBossRuntime.java:31)
	at ceylon.modules.jboss.runtime.AbstractJBossRuntime.createClassLoader(AbstractJBossRuntime.java:78)
	at ceylon.modules.api.runtime.AbstractRuntime.execute(AbstractRuntime.java:104)
	at ceylon.modules.Main.execute(Main.java:69)
	at ceylon.modules.Main.main(Main.java:42)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(java.base@9.0/Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(java.base@9.0/NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(java.base@9.0/DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(java.base@9.0/Method.java:531)
	at org.jboss.modules.Module.run(Module.java:308)
	at org.jboss.modules.Main.main(Main.java:487)
```

I had to add:

```
-XaddExports:java.xml/com.sun.org.apache.xerces.internal.jaxp=ALL-UNNAMED,java.xml/com.sun.org.apache.xalan.internal.xsltc.trax=ALL-UNNAMED,java.xml/com.sun.xml.internal.stream=ALL-UNNAMED
```

To the VM options to make it work. Note that I'm using `Main.main` so moving the static init there explicitly would not help me.

Can you explain what this is trying to fix? Is it really specific to WildFly? Or is it something that we'll regret taking out?

Thanks, cheers.

**Comment by dmlloyd** [comment]
@fromage commented

So yeah, OK the exception I got is not the same as this issue's title, but the cause and possibly fix should be the same :)

**Comment by dmlloyd** [comment]
@dmlloyd commented

I can say two directly pertinent things about this.

1. The redirect packages will not be needed in Java 9 if we get modularity there, as the problems they are meant to solve should be fixed there.
2. Jigsaw is not Java 9, has not been merged, and probably won't be merged (as is).

**Comment by dmlloyd** [comment]
@dmlloyd commented

As to why the redirect packages are necessary - it is presently impossible to establish a default JAXP implementation in the JDK (8 and earlier) that is not in the application class loader because it uses a system property to do so.  So the redirect packages essentially act as the JDK's idea of the default JAXP implementation (because jboss-modules.jar is on the application class loader), which then in turn hold references to the "real" implementation that can then be loaded from modules or wherever.

**Comment by dmlloyd** [comment]
@fromage commented

So you're saying we _need_ `__redirected` then, in Java 8, right? Even outside of WildFly?

**Comment by dmlloyd** [comment]
@dmlloyd commented

Only if you want JAXP to work correctly.  Though I think if you stick with the JDK implementation, it may work OK without it.

**Comment by dmlloyd** [comment]
**[George Gastaldi](https://issues.redhat.com/secure/ViewProfile.jspa?name=gastaldi)** commented

We have a JBoss Modules fork because of this JAXP issue which is hard to maintain, so we would love to use the official JAR. Can this be resolved by introducing a system property to avoid calling https://github.com/jboss-modules/jboss-modules/blob/master/src/main/java/org/jboss/modules/Module.java#L133?

**Comment by dmlloyd** [comment]
**[Gunnar Morling](https://issues.redhat.com/secure/ViewProfile.jspa?name=JIRAUSER200519)** commented

This is hitting us in the Maven [plug-in for the patch-gen tool](https://github.com/jbossas/patch-gen/tree/master/patch-gen-maven-plugin), too. The redirection of `SAXParserFactory` et al. causes plug-ins running later in the build process to fail as they don't have the redirected classes on their plug-in path.

**Comment by dmlloyd** [comment]
@dmlloyd commented

This should be fixed as of 1.8.0.Final.

**Comment by dmlloyd** [comment]
**[Lincoln Baxter III](https://issues.redhat.com/secure/ViewProfile.jspa?name=lincolnthree)** commented

Woooo! Nice [dlloyd@redhat.com](https://issues.redhat.com/secure/ViewProfile.jspa?name=dlloyd@redhat.com) :)
